spring-expression/src/main/java/org/springframework/expression/spel/ast/Indexer.java

```diff
@@ -187,11 +190,73 @@ else if (target instanceof Collection<?> collection) {
 			return new PropertyIndexingValueRef(
 					target, (String) index, state.getEvaluationContext(), targetDescriptor);
 		}
-
+		Optional<ValueRef> optional = tryIndexAccessor(state, index);
+		if (optional.isPresent()) {
+			return optional.get();
+		}
 		throw new SpelEvaluationException(
 				getStartPosition(), SpelMessage.INDEXING_NOT_SUPPORTED_FOR_TYPE, targetDescriptor);
 	}
 
+	private Optional<ValueRef> tryIndexAccessor(ExpressionState state, Object index) {
+		EvaluationContext context = state.getEvaluationContext();
+		Object target = state.getActiveContextObject().getValue();
+		if (context != null) {
+			List<IndexAccessor> list = context.getIndexAccessors();
+			if (list != null) {
+				List<IndexAccessor> availableAccessors = getIndexAccessorsToTry(state.getActiveContextObject(), list);
+				try {
+					for (IndexAccessor indexAccessor : availableAccessors) {
+						if (indexAccessor.canRead(context, target, index)) {
+							ValueRef valueRef = indexAccessor.read(context, target, index);
```

Comment:
The `ValueRef` will need to be a local implementation since `IndexAccessor#read` must return a `TypedValue`.